### ISSUE TITLE
Meta: Including non-dark theme tweaks, e.g hiding tabs when using treestyle

### ISSUE THREAD

**Issue by krruzic** [overdodactyl/ShadowFox#53]
Loving this theme, had made a similar one but this fixes a lot of my issues. Anyway, was wondering if/how CSS could be used if you're using treestyle tabs that hides the whole tab bar.

I've put together some that does this but no idea how to merge it in without forcing it on people.
```css
/* hide tabs */
#tabbrowser-tabs { visibility: hidden !important; }
tab { display: none !important; }

/* CSS to make urlbar and window controls 1 line */
#nav-bar {      
  margin-bottom: -1px !important;               /* remove navbar bottom 1px border */
  margin-top: -30px !important;
  margin-right: 7.2vw !important; 
  border-top: none !important;
}   

/* Remove padding above tabbar in compact mode */
#main-window[sizemode="normal"] > #titlebar {
  -moz-appearance: initial !important;
}

/* Back & Forward buttons */  
#back-button > .toolbarbutton-icon,
#forward-button > .toolbarbutton-icon {
  transform: scale(0.9, 0.9) !important;
  margin-left: -2px !important;
  margin-right: -2px !important; 
}

/* Move hamburger menu to the left */
#PanelUI-button,
#customization-panel-container,
#customization-panelWrapper .panel-arrow { 
  -moz-box-ordinal-group: 0 !important;
  margin-right: -6px !important;
}

#appMenu-popup {
  margin-right: -258px !important;
}

#appMenu-popup .panel-arrow {
  margin-right: 248px !important;
}

/* More tools... button */
#nav-bar-overflow-button { 
  transform: scale(0.9, 0.9) !important;
}

#titlebar-buttonbox {
  background: var(--chrome-secondary-background-color) !important;
}
```

Ends up looking like this: 
![oneline](https://user-images.githubusercontent.com/4000854/34647413-6499f1bc-f33f-11e7-8074-032b0da81009.PNG)

**Comment by overdodactyl** [comment]
Hi @krruzic,

Thanks for sharing your code, I like the look!

I've been trying to decide what should be done with snippets like this that begin to stray away from being about a dark theme, but obviously could be of great use to some users.

I have several of my own changes that might be nice to include, but I'm not sure of the best approach to take.  

My initial feelings are that significant changes to the UI shouldn't be included in the repo by default (as to avoid forcing changes on people as you mentioned).  

Given that, two approaches could be taken:

* Include extra folders for userChrome/Content tweaks that aren't specific to a dark theme.  `userChrome_imports.css` and `userContent_imports.css` would have the necessary `import` statements commented out by default.  Users using `gulp` would have to add in the necessary commonds to include extra tweaks, and those using the default `user*.css` would have to copy/paste the snippets they want.

* Refer people to another repository with collections of tweaks.  If this approach was taken, I could either set up my own that's more closely related to this repo (things that correlate with extensions I've modified - such as what you've provided here, more theme related modifications such as changing icon colors, favicons, png's used for bookmarks etc.).  Alternatively, I could direct people to more broad/inclusive repositories that are already fairly well established.  For example: https://github.com/Timvde/UserChrome-Tweaks



Anyone have any thoughts?

**Comment by CaptaPraelium** [comment]
TL;DR Everything in one place is good, but mixing projects is not.

I have a few tweaks I've added, to bring back pre-57 features that I like, but they are definitely not shadowfox-related (Auto-hiding close buttons on every tab, non-auto-hiding 'list all tabs' button). I was working on autohiding the toolbars since that can no longer be done with a keyboard shortcut via an addon (at least, until the new APIs are released...grrr Mozilla!) While these are really useful to me, I'd never consider adding them to shadowfox. I'm from the linux thinking of 'do one job and do it well', and to me, shadowfox's job is 'make stuff dark'.

Personally, I feel that all of these kind of tweaks should be in one place. It was rather annoying to search the web for CSS, only to find that much of it was outdated or buggy, and have to keep searching, mixing parts of different scripts, etc etc.

Not to mention, since you linked that repo, I've found more tweaks that I think I might use - thanks for that! My point though, is that if everything is in one place, it also makes more possibilities apparent to the end-user (I never even considered greyscale favicons in tabs, but I've just seen that tweak and I'm like 'hmmm sounds nice! I wonder if I could do half-size favicons, too?')

**Comment by overdodactyl** [comment]
>Everything in one place is good, but mixing projects is not.

Ya these were my thoughts as well.  I have a lot of stuff I could share, but definitely isn't dark theme related (including the autohiding toolbars if you're still in need of that)...and hadn't really thought about adding them to this repo.

Having everything together is also nice though, which is the only reason I really was considering adding stuff to the repo.  Are you of the opinion this should just stay dark-theme related then and perhaps put together a wiki entry of available resources?  Or were you thinking of trying to make this an all in one repo?

**Comment by krruzic** [comment]
thanks for the feedback. I agree, just thought something to hide the tabs would be useful as this theme does include treestyle modifications.

**Comment by CaptaPraelium** [comment]
Actually, I think a wiki is an outstanding idea, and there's one strong reason for that. In the repo above, I found a LOT of bugs. I can go through and fix them up but the process of doing so via git is overly complicated. And, what if my code leaves bugs or introduces new ones? Then someone else has to go through that same hassle. More time would end up being spent on maintaining the availability of the code, than actually writing it. A wiki makes it simple, and fast.

The downside of a wiki I guess, is that the files are not directly downloadable, but, I think that when it comes to these kind of tweaks, a user is going to want to mix together their own collection of tweaks manually. The only real alternatives, we have been through before - buggy/inconsistent @import statements, or somewhat complex apps to blend things together.... and again we come back to an issue of high maintenance. I think copy/pasting from a wiki works just fine.

BTW, one example of such bugs, and of specific use-cases, is autohiding toolbars. On windows, the window manager buttons (minimise, maximise/restore, close) are part of the tab bar - there is no separate toolbar here. Using scripts for resizing and hiding toolbars which I have found online, invariably fails as a result of these titlebar buttons. Even for those of us who would use something like treestyletabs, which I don't because I use this as a video player and I want the horizontal real estate, can't hide that toolbar as it means hiding the window management. We can autohide, and then it makes parts of the webpages inaccessible.... And of course that brings me back to the original issue of existing tweaks not working properly on windows due to it's unique handling of window management................I could go on, but I think you get my point here.

**Comment by overdodactyl** [comment]
>just thought something to hide the tabs would be useful as this theme does include treestyle modifications.

I agree! Just trying to decide the best way way to go about doing so :)

>Actually, I think a wiki is an outstanding idea

I was originally thinking of just a wiki entry with links to outside resources, but I could definitely do a wiki with code snippets included.  Something like what's provided on the Tree Style Tab wiki page comes to mind:

https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules

>The downside of a wiki I guess, is that the files are not directly downloadable

The only other downsides I can think of would be:

* Not as easy for users to make contributions (wiki's aren't included when you fork a repo using GitHub online)...so for a user to "submit" code, they would have to either use `git` to clone the wiki (separate from cloning the main repo) or open up an issue.  But, perhaps this isn't a huge issue as I've only had a few pull requests submitted anyways...

* Possibly a little more upkeep for me (if code has to be changed, it's easier to just `git push` than update a wiki.  But these types of tweaks might not have to be changed that often, so perhaps that's not a huge issue.  However, initial set-up might take a little longer, especially if it's well done, but I think the following pluses could make it worth that

Additional benefits:

* Additional tweaks are "out of the way" of the main repo.  In other words, the main focus of the ShadowFox code stays dark-theme related, but could add all sorts of extra helpful things like @krruzic posted could still be included

* Could easily add photos and descriptions for what each tweak does/looks like

**Comment by CaptaPraelium** [comment]
Sorry, in my thinking, a wiki meant it is publicly editable. That was the real drawcard for me. Ease of contribution is really paramount. I can see now that github wikis don't really work that way. I was thinking more along the lines of a wikipedia page, where anyone can edit and there is a history of edits. I wonder if there's another online service which would provide such an interface?

**Comment by overdodactyl** [comment]
I'm not aware of any other online services like that off the top of my head.  

I'm curious if something like that did exist though, if it would garnish much participation.  My concerns would be that since the other repo already exists, there would be a lot of redundancy and not a ton of interest.  In which case it would probably be best to keep things here.

Issues aren't the best/easiest way to submit code or changes, but it could work (possibly), as could a GitHub repo.

In the cases where people don't want to use `git`, github files can be added/edited directly from the online interface.

**Comment by overdodactyl** [comment]
Another possibility would be just using a `.md` file (such as README.md, which is displayed on the home page of this repo) instead of the wiki.  This would let people edit it and would be pretty much the same as the wiki, just not available from the wiki tab.

**Comment by overdodactyl** [comment]
Hmmm, maybe you can make wiki's public...

![public_wiki](https://user-images.githubusercontent.com/33071129/34755607-f56b30fa-f5bc-11e7-8a07-b6bb1d2704c3.png)

**Comment by CaptaPraelium** [comment]
Hmm, what is a "contributor"? I mean if you could make it editable to a select group of users, that might help....

The .md files seem like a workable solution too... but it seems like I need write access to the project, to edit them... And that looks like I need to be a "contributor" too. Seems kinda 'all or nothing'? :(

**Comment by overdodactyl** [comment]
Do you mean collaborator or contributor?  I think anyone who makes a pull request that gets merged automatically gets the label "contributor".

Collobaroters on the other hand get special permissions on the repo.  

> Collaborators on a personal repository can:
> 
>     - Push to (write), pull from (read), and fork (copy) the repository
>     - Create, apply, and delete labels and milestones
>     - Open, close, re-open, and assign issues
>     - Edit and delete comments on commits, pull requests, and issues
>     - Mark an issue or pull request as a duplicate. For more information, see "About duplicate issues and pull requests."
>     - Merge and close pull requests
>     - Send pull requests from forks of the repository
>     - Create and edit Wikis
>     - Create and edit Releases
>     - Remove themselves as collaborators on the repository
>     - Submit a review on a pull request that will affect its mergeability
>     - Act as a designated code owner for the repository. For more information, see "About CODEOWNERS."
>     * Lock a conversation. For more information, see "Locking conversations."

https://help.github.com/articles/permission-levels-for-a-user-account-repository/

To edit a .md file as a non-collaborator, I think you would have to fork the repo, make the changes, then submit a PR, which can all be done via the web interface, but not everyone will want to do that or necessarily know how.

**Comment by CaptaPraelium** [comment]
LOL, helps if I L2read XD sorry about mixing up the two there.
Yeh that was what I saw, it automatically forked it for me. I think that's a bit heavy when someone might just want to add one line of text to a code snippet. I mean, I use git and svn at work, but for home fun stuff like adding a windows-specific fix to some CSS that makes my favicons dark mode, I just want to 'click edit, paste, click save.'

I think it's probably worth making a new repo, and making that wiki publicly editable, so you don't have to open your documentation on this repo to potential trouble.

**Comment by overdodactyl** [comment]
At the same time though it would nice to keep everything contained in this repo.  I would hope no one would go through the wiki with the intent of changing things for the worse.  I think it's worth opening up the wiki here for a trial run (everything's backed up in case something goes wrong).

I've just very quickly thrown together an example page:
https://github.com/overdodactyl/ShadowFox/wiki/More-Tweaks

Would you mind trying to edit it with something at some point? My hope is that I would get a notification or something letting me know someone revised it

**Comment by CaptaPraelium** [comment]
Yeh that's annoying isn't it. I mean, on the one hand, it's just a small community so you want to remove any barriers to contribution, and on the other hand, those barriers keep us safe.

I see the wiki has page history, so hopefully restoring shouldn't be a big problem if the worst should happen.

So, did you get a notification?

**Comment by overdodactyl** [comment]
Thanks for the additions!

Nope, no notifications unfortunately.  

Looking into it a bit, looks like GitHub doesn't support that.  Appears [StackOverflow](https://stackoverflow.com/questions/8407917/how-can-you-track-or-be-notified-of-changes-to-github-wikis) suggestions include either using a RSS feeder or service hooks...so that could work.  

I'd like to decrease as many barriers to contributing as possible...so I think this will be a good way to try until (hopefully never) someone gives me a reason to close it again haha.

I guess this introduces a need to put together a new CONTRIBUTING.md file (which I've been meaning to do anyways).  I noticed you gave code attribution...but that's definitely something that should be emphasized.

**Comment by overdodactyl** [comment]
@krruzic, I've added your setup to the wiki: https://github.com/overdodactyl/ShadowFox/wiki/More-Tweaks#combined-setups

Feel free to change/edit it as you see fit! Thanks again :)
